test/e2e/framework/resource/resources.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/onsi/ginkgo"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	scaleclient "k8s.io/client-go/scale"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	testutils "k8s.io/kubernetes/test/utils"
+)
+
+const (
+	// Number of objects that gc can delete in a second.
+	// GC issues 2 requestes for single delete.
+	gcThroughput = 10
+)
+
+// ScaleResource scales resource to the given size.
+func ScaleResource(
+	clientset clientset.Interface,
+	scalesGetter scaleclient.ScalesGetter,
+	ns, name string,
+	size uint,
+	wait bool,
+	kind schema.GroupKind,
+	gvr schema.GroupVersionResource,
+) error {
+	ginkgo.By(fmt.Sprintf("Scaling %v %s in namespace %s to %d", kind, name, ns, size))
+	if err := testutils.ScaleResourceWithRetries(scalesGetter, ns, name, size, gvr); err != nil {
+		return fmt.Errorf("error while scaling RC %s to %d replicas: %v", name, size, err)
+	}
+	if !wait {
+		return nil
+	}
+	return e2epod.WaitForControlledPodsRunning(clientset, ns, name, kind)
```

Comment:
@gavinfish i think this is where you are getting a circular dependecy (?)
Given the struct of `e2epod.WaitForControlledPodsRunning()` in https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pod/wait.go#L432 I think it would be best to move that function to this subpkg (e2eresource).

`ScaleResource` feels pretty general (better to be in the e2eresource than e2epod).
And the way `e2epod.WaitForControlledPodsRunning` is structure fits in nicely with this subpkg.

Reply:
Agree. And then `e2epod.WaitForControlledPods()` also needs to be moved to e2eresource.